[Merged by Bors] - feat(RingTheory/MvPolynomial/{MonomialOrder,Ideal}): leadingTerm

tsuki8 opened the pull request:
define the leadingTerm and prove some lemmas related to the def

`leadingTerm`: the leading term of `f` for the monomial ordering `m`

some basic lemmas about leadingTerm including:
1. `leadingTerm_eq_zero_iff`
2. `image_leadingTerm_sdiff_singleton_zero`
3. `image_leadingTerm_insert_zero`
4. `leadingTerm_zero`
5. `degree_leadingTerm`
6. `leadingCoeff_leadingTerm`
7. `leadingTerm_leadingTerm`
8. `leadingTerm_C`

some lemmas about the degree of `f - m.leadingTerm f`:
1. `degree_sub_leadingTerm_le`
2. `degree_sub_leadingTerm_lt_degree`
3. `degree_sub_leadingTerm_lt_iff`
4. `degree_ne_zero_of_sub_leadingTerm_ne_zero`

also some lemmas about leading terms and ideals：
1. `span_leadingTerm_sdiff_singleton_zero`
2. `span_leadingTerm_insert_zero`
3. `span_leadingTerm_eq_span_monomial`
4. `span_leadingTerm_eq_span_monomial'`
5. `span_leadingTerm_eq_span_monomial₀` 

Co-authored-by: Junyu Guo @Hagb <hagb@hagb.name>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

- [x] depends on: #24361
- [x] depends on: #26148

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Hagb committed:
feat(Mathlib/Data/Finsupp/MonomialOrder): `0 ≤ a` where `a : m.syn`

`simp` would be able to solve it.

Hagb committed:
feat(RingTheory/Ideal/Span.lean): some lemmas about zero and span

`simp` would be able to solve them.

Notice that `span_singleton_zero : span {(0 : R)} = ⊥` itself can be
solved by `simp` without this commit, which uses
`span_singleton_eq_bot : span ({x} : Set α) = ⊥ ↔ x = 0`, while
`simp` cannot solve statements like `f (Ideal.span {(0 : R)}) = f ⊥`,
which would be solvable by `simp` with `span_singleton_zero` marked
with `@[simp]`.

Hagb committed:
Update Mathlib/RingTheory/DedekindDomain/Ideal.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/DedekindDomain/Ideal.lean

Co-authored-by: Thomas Browning <tb65536@users.noreply.github.com>

Hagb committed:
move `dvd_bot` to where it should be

Hagb committed:
feat(Mathlib/RingTheory/MvPolynomial/MonomialOrder): some lemmas about degree

tsuki8 committed:
Merge remote-tracking branch 'origin/mv-degree' into leadingTerm
The theorem need to use the pr submitted by Hagb

tsuki8 committed:
Merge remote-tracking branch 'upstream/MonomialOrder-zero_le' into leadingTerm
need to use zero_le

tsuki8 committed:
feat(RingTheory/MvPolynomial/MonomialOrder): leadingTerm
define the leadingTerm and prove some lemmas related to the def

`leadingTerm`: the leading term of `f` for the monomial ordering `m`

some basic lemmas about leadingTerm including:
1. `leadingTerm_eq_zero_iff`
2. `leadingTerm_image_sdiff_singleton_zero`
3. `leadingTerm_image_insert_zero`
4. `leadingTerm_zero`
5. `leadingTerm_degree_eq`
6. `leadingTerm_degree_eq'`

some lemmas about the degree of `f - m.leadingTerm f`:
1. `degree_sub_leadingTerm`
2. `degree_sub_leadingTerm_lt_degree`
3. `degree_sub_leadingTerm_lt_iff`

tsuki8 committed:
submit `leadingTerm` and related lemmas

tsuki8 committed:
change based on the github actions

Hagb committed:
add `degree_sum_le`

Hagb committed:
`lemma` -> `theorem`

Hagb committed:
Merge branch 'master' into Ideal-zeros

Hagb committed:
Update Span.lean

Hagb commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -636,6 +733,65 @@ theorem leadingCoeff_sub_of_lt {f g : MvPolynomial σ R} (h : m.degree g ≺[m]
+        refine mem_support_iff.mp ?_
+        exact hin
+      exact this hc
+    have h₃: m.toSyn (m.degree (f - m.leadingTerm f)) ≤  m.toSyn (m.degree f) := by
+      have h₃': m.toSyn (m.degree (f - m.leadingTerm f)) ≤
+          m.toSyn (m.degree f) ⊔ m.toSyn (m.degree (m.leadingTerm f)) := by
+        apply degree_sub_le
+      have h₃'':  m.toSyn (m.degree f) = m.toSyn (m.degree (m.leadingTerm f)) := by
+        exact Eq.symm (leadingTerm_degree_eq' f)
+      have h3:  m.toSyn (m.degree f) ⊔ m.toSyn (m.degree (m.leadingTerm f)) =
```
```suggestion
    have h₃ : m.toSyn (m.degree (f - m.leadingTerm f)) ≤  m.toSyn (m.degree f) := by
      have h₃' : m.toSyn (m.degree (f - m.leadingTerm f)) ≤
          m.toSyn (m.degree f) ⊔ m.toSyn (m.degree (m.leadingTerm f)) := by
        apply degree_sub_le
      have h₃'' : m.toSyn (m.degree f) = m.toSyn (m.degree (m.leadingTerm f)) := by
        exact Eq.symm (leadingTerm_degree_eq' f)
      have h3 : m.toSyn (m.degree f) ⊔ m.toSyn (m.degree (m.leadingTerm f)) =
```

Hagb commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -601,6 +638,66 @@ protected theorem Monic.prod {ι : Type*} {P : ι → MvPolynomial σ R} {s : Fi
+  rw [leadingTerm_eq_zero_iff]
+
+/-- The degree of `f` equals to the degree of `leadingTerm f` -/
+lemma leadingTerm_degree_eq (f : MvPolynomial σ R) :
+  m.degree (m.leadingTerm f) = m.degree f := by
+    classical
+    by_cases h : f = 0 <;> simp [leadingTerm,h]
+    have : m.leadingCoeff f != 0 := by
+      simp [leadingCoeff, h]
+    simp [MonomialOrder.degree_monomial]
+    exact fun a ↦ False.elim (h a)
```
```suggestion
lemma leadingTerm_degree_eq (f : MvPolynomial σ R) :
    m.degree (m.leadingTerm f) = m.degree f := by
  classical
  by_cases h : f = 0 <;> simp [leadingTerm,h]
  have : m.leadingCoeff f != 0 := by
    simp [leadingCoeff, h]
  simp [MonomialOrder.degree_monomial]
  exact fun a ↦ False.elim (h a)
```

Hagb commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -601,6 +638,66 @@ protected theorem Monic.prod {ι : Type*} {P : ι → MvPolynomial σ R} {s : Fi
+    simp [MonomialOrder.degree_monomial]
+    exact fun a ↦ False.elim (h a)
+
+lemma leadingTerm_degree_eq' (f : MvPolynomial σ R) :
+  m.toSyn (m.degree (m.leadingTerm f)) = m.toSyn (m.degree f) := by
+    classical
+    by_cases h : f = 0 <;> simp [leadingTerm,h]
+    have : m.leadingCoeff f != 0 := by
+      simp [leadingCoeff, h]
+    simp [MonomialOrder.degree_monomial]
+    exact fun a ↦ False.elim (h a)
```
```suggestion
lemma leadingTerm_degree_eq' (f : MvPolynomial σ R) :
    m.toSyn (m.degree (m.leadingTerm f)) = m.toSyn (m.degree f) := by
  classical
  by_cases h : f = 0 <;> simp [leadingTerm,h]
  have : m.leadingCoeff f != 0 := by
    simp [leadingCoeff, h]
  simp [MonomialOrder.degree_monomial]
  exact fun a ↦ False.elim (h a)
```

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

tsuki8 committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Hagb (Junyu Guo 郭俊余) <hagbgreen@yahoo.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Ruben Van de Velde <65514131+Ruben-VandeVelde@users.noreply.github.com>

Hagb committed:
remove unnecessary `classical`

Hagb committed:
Merge commit '4449ce7cca436f2399ffe61ad1546741862144ac' into HEAD

Hagb committed:
improve naming and some proofs

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
fix style

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean

Co-authored-by: Riccardo Brasca <riccardo.brasca@gmail.com>

Hagb committed:
Merge remote-tracking branch 'origin/Ideal-zeros' into HEAD

Hagb committed:
Merge remote-tracking branch 'origin/mv-degree' into HEAD

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

chrisflav reviewed (commented):
Thanks for your PR! I left a first round of comments below.

Hagb committed:
Apply suggestions from code review

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Apply suggestions from code review

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Merge remote-tracking branch 'upstream/master' into HEAD

Hagb committed:
add lemmas about ideal

Hagb committed:
squeeze non-terminal `simp`

Hagb committed:
move `leadingTerm_zero` right after `leadingTerm_zero_iff`

Hagb commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -605,6 +620,58 @@ protected theorem Monic.prod {ι : Type*} {P : ι → MvPolynomial σ R} {s : Fi
+of all polynomials in B, excluding zero.
+-/
+lemma image_leadingTerm_sdiff_singleton_zero (B : Set (MvPolynomial σ R)) :
+    m.leadingTerm '' (B \ {0}) = (m.leadingTerm '' B) \ {0} := by
```
Thanks for your review and suggestion! And sorry for our delay.

It is for the following proposition about Gröbner basis, which we are submitting to Mathlib4 in #29203 from https://github.com/WuProver/groebner_proj.

https://github.com/leanprover-community/mathlib4/blob/508ec073409f1dc879f98a21f8dcefde62a5ccf0/Mathlib/RingTheory/MvPolynomial/Groebner.lean#L637-L641

`span_leadingTerm_sdiff_singleton_zero`, submitted in `Mathlib/RingTheory/MvPolynomial/Ideal.lean` in this PR, uses `image_leadingTerm_sdiff_singleton_zero` directly:

```lean
lemma span_leadingTerm_sdiff_singleton_zero (B : Set (MvPolynomial σ R)) :
    span (m.leadingTerm '' (B \ {0})) = span (m.leadingTerm '' B) :=
  m.image_leadingTerm_sdiff_singleton_zero B ▸ Ideal.span_sdiff_singleton_zero
```

`IsGroebnerBasis` is defined as following. https://github.com/leanprover-community/mathlib4/blob/508ec073409f1dc879f98a21f8dcefde62a5ccf0/Mathlib/RingTheory/MvPolynomial/Groebner.lean#L626-L635

(There are also a few miscellaneous theorems dependent on it, referring to [the dependency graph of our project](https://wuprover.github.io/groebner_proj/blueprint/dep_graph_document.html).)

Hagb replied:
There are another two proofs using it (also in #29203):
https://github.com/leanprover-community/mathlib4/blob/0a27a6eb4d2a1183067ff5e1a84c590643f321b0/Mathlib/RingTheory/MvPolynomial/Groebner.lean#L620-L630
https://github.com/leanprover-community/mathlib4/blob/0a27a6eb4d2a1183067ff5e1a84c590643f321b0/Mathlib/RingTheory/MvPolynomial/Groebner.lean#L692-L697

Hagb replied:
> They are not using `image_leadingTerm_sdiff_singleton_zero` but the span version right? I am asking if `image_leadingTerm_sdiff_singleton_zero` is used anywhere else.

Hmmm I confused myself. Sorry about that.

There are two in #29203:

https://github.com/leanprover-community/mathlib4/blob/0a27a6eb4d2a1183067ff5e1a84c590643f321b0/Mathlib/RingTheory/MvPolynomial/Groebner.lean#L1265-L1273

https://github.com/leanprover-community/mathlib4/blob/0a27a6eb4d2a1183067ff5e1a84c590643f321b0/Mathlib/RingTheory/MvPolynomial/Groebner.lean#L1324-L1330

Hagb committed:
improve the document of `degree_sub_leadingTerm_lt_iff`

Hagb committed:
Merge branch 'master' into leadingTerm

kim-em commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -638,6 +695,38 @@ theorem leadingCoeff_sub_of_lt {f g : MvPolynomial σ R} (h : m.degree g ≺[m]
+    m.degree (f - m.leadingTerm f) ≺[m] m.degree f ↔ m.degree f ≠ 0 := by
+  constructor
+  · intro h h'
+    -- simp? [h'] at h
```
Cleanup?

Hagb replied:
Makes sense. I've cleaned these comments up.

Hagb committed:
try to resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2472722695

Hagb committed:
try to resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2472728107

chrisflav commented on Mathlib/RingTheory/MvPolynomial/Ideal.lean:
```diff
@@ -50,3 +51,70 @@ theorem mem_ideal_span_X_image {x : MvPolynomial σ R} {s : Set σ} :
+    simp_rw [MonomialOrder.leadingTerm, ← submodule_span_eq,
+      Submodule.mem_span_image_iff_exists_fun] at *
+    rcases hl with ⟨t, ht, c, hc⟩
+    rw [←hc]
```
```suggestion
    rw [← hc]
```
See https://leanprover-community.github.io/contribute/style.html#whitespace-and-delimiters

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Christian Merten <christian@merten.dev>

Hagb committed:
remove `degree_sub_leadingTerm{,_lt_degree}` from module doc (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2486569590)

Hagb committed:
remove comment `-- simp?` if the unsqueezed `simp` is short (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2486679733)

chrisflav committed:
simplify `span_leadingTerm_eq_span_monomial` and add `C_mul_leadingCoeff_monomial_degree` (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2486951025)

erdOne commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -108,6 +113,16 @@ if its leading coefficient (for that monomial order) is 1. -/
 def Monic (f : MvPolynomial σ R) : Prop :=
   m.leadingCoeff f = 1
 
+/-- The leading term of of a multivariate polynomial with respect to a monomial ordering. -/
```
```suggestion
variable (m) in
/-- The leading term of of a multivariate polynomial with respect to a monomial ordering. -/
```

Hagb committed:
generate `degree_sub_leadingTerm_lt_degree` with some lemmas added

resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2529881858 and https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2529879468

Hagb committed:
remove `degree_sub_leadingTerm` (resolve https://github.com/leanprover-community/mathlib4/pull/26039#discussion_r2529882329)

Hagb committed:
Apply suggestion from @erdOne

Co-authored-by: Andrew Yang <36414270+erdOne@users.noreply.github.com>

Hagb commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -638,6 +694,42 @@ theorem leadingCoeff_sub_of_lt {f g : MvPolynomial σ R} (h : m.degree g ≺[m]
+  rw [h', mem_support_iff] at h
+  simp [leadingTerm, leadingCoeff] at h
+
+theorem degree_sub_leadingTerm_lt_degree {f : MvPolynomial σ R} (h : f - m.leadingTerm f ≠ 0) :
```
Thanks! Resolved in [`b8c68bf`](https://github.com/leanprover-community/mathlib4/pull/26039/commits/b8c68bfefb8ac0ddd1cd4f87219f3453056dda6e).

Hagb commented on Mathlib/RingTheory/MvPolynomial/MonomialOrder.lean:
```diff
@@ -640,6 +707,37 @@ theorem leadingCoeff_sub_of_lt {f g : MvPolynomial σ R} (h : m.degree g ≺[m]
+  simp [degree_leadingTerm]
+
+theorem degree_sub_leadingTerm (f : MvPolynomial σ R) :
+    m.degree (f - m.leadingTerm f) ≺[m] m.degree f ∨ f - m.leadingTerm f = 0 := by
```
> I would be keen to see this gone though. Especially because we have `degree_sub_leadingTerm_lt_iff`.

@erdOne Okay. It has been removed in [`d564a65`](https://github.com/leanprover-community/mathlib4/pull/26039/commits/d564a6569b1537ba8fed4f4ae38793bae7abeb3e).

Hagb committed:
Update Mathlib/RingTheory/MvPolynomial/Ideal.lean

Co-authored-by: Dagur Asgeirsson <dagurtomas@gmail.com>

Hagb committed:
Merge branch 'master' into leadingTerm

erdOne commented:
Can you make sure if the PR description is up to date?
Otherwise LGTM. Thanks!
maintainer delegate
